File: leetcode/delete-node-in-a-bst/python/delete-node-in-a-bst.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
def findMaxValue(root):
    current = root
    while current and current.right: # current는 current.right이 유효하기 위한 조건
        current = current.right
    return current


class Solution:
    def deleteNode(self, root, key):
        # base condition
        if not root:
            return None

        if key < root.val:
            root.left = self.deleteNode(root.left, key)
        elif root.val < key:
            root.right = self.deleteNode(root.right, key)
        else:  # key == root.val
            if not root.left:
                return root.right
            elif not root.right:
                return root.left
            else:
                # find the max value int left hand side
                max_node = findMaxValue(root.left)
                root.val = max_node.val  # 나보다 작지만 그 중에선 제일 큰 걸로 변경 -> BST 규칙 유지됨
                root.left = self.deleteNode(root.left, max_node.val)  # 중복 없애야 하므로 삭제

        return root

        

File: leetcode/delete-node-in-a-bst/c/delete-node-in-a-bst.c
#include <stdlib.h>

// Definition for a binary tree node.
// struct TreeNode {
//     int val;
//     struct TreeNode *left;
//     struct TreeNode *right;
// };

// Helper function to find the minimum node in the right subtree
struct TreeNode* findMinNode(struct TreeNode* root) {
    struct TreeNode* current = root;
    while (current && current->left) {
        current = current->left;
    }
    return current;
}

struct TreeNode* deleteNode(struct TreeNode* root, int key) {
    if (!root) {
        return NULL;
    }
    
    // Find the node to delete
    if (key < root->val) {
        root->left = deleteNode(root->left, key);
    } else if (key > root->val) {
        root->right = deleteNode(root->right, key);
    } else {
        // Node to be deleted is found

        // Node with only one child or no child
        if (root->left == NULL) {
            struct TreeNode* temp = root->right;
            free(root);
            return temp;
        } else if (root->right == NULL) {
            struct TreeNode* temp = root->left;
            free(root);
            return temp;
        } else {
            // Node with two children: Get the inorder successor (smallest in the right subtree)
            struct TreeNode* temp = findMinNode(root->right);

            // Copy the inorder successor's value to this node
            root->val = temp->val;

            // Delete the inorder successor
            root->right = deleteNode(root->right, temp->val);
        }
    }

    return root;
}
